### PR TITLE
NF(TST): Creating tests for misc.py

### PR DESCRIPTION
**Pull request by glalteva**

### CODE REVIEW

File: datalad/crawler/nodes/misc.py [hunk 1]
```diff
@@ -56,14 +57,19 @@ def get_values(self, *keys):
...
     def __call__(self, data):
-        # ??? for some reason didn't work when I made entire thing a list
         if self.keys:
-            raise NotImplementedError("Jason will do it")
+            for k in data.keys():
+                size = len(self.keys)
+                for i in range(0, size):
+                    key = self.keys[i]
+                    if key == k:
+                        self.data.append({key: data.get(key)})
```
Comment (yarikoptic) [inline datalad/crawler/nodes/misc.py]: Let's work on this piece... 
Python is great! You could do just
`for key in self.keys:`
To traverse the list of keys. Adjust the code with that in mind. Then note my note I will leave in the unittest

File: datalad/crawler/nodes/tests/test_misc.py [hunk 2]
```diff
@@ -47,6 +48,23 @@ def test_get_disposition_filename():
...
+    genempty = Sink(None, None)
+    eq_(list(genempty(data)), [{'a': 'b', 'x': 'y', 'g': 'h'}])
+
+    # if key for sunk data is specified
+    gen = Sink(None, 'result')
+    eq_(list(gen(data)), [{'a': 'b', 'x': 'y', 'result': [{'a': 'b', 'x': 'y', 'g': 'h'}], 'g': 'h'}])
+
+    # if list of keys is specified
+    genkeys = Sink(keys, 'result')
+    eq_(list(genkeys(data)), [{'a': 'b', 'x': 'y', 'result': [{'a': 'b'}, {'x': 'y'}], 'g': 'h'}])
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: In result, x and a entries should be a part of the same dict, not a list of dicts

File: datalad/crawler/nodes/tests/test_misc.py [hunk 3]
```diff
@@ -47,6 +48,23 @@ def test_get_disposition_filename():
     eq_(output[0]['filename'], 'T1.nii.gz')
 
 
+def test_sink():
+    data = {'x': 'y', 'g': 'h', 'a': 'b'}
+    keys = ['x', 'a']
+
+    # no arguments
+    genempty = Sink(None, None)
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: No need to specify args, identical to defaults

File: datalad/crawler/nodes/tests/test_misc.py [hunk 4]
```diff
@@ -47,6 +48,23 @@ def test_get_disposition_filename():
...
+    eq_(list(genempty(data)), [{'a': 'b', 'x': 'y', 'g': 'h'}])
+
+    # if key for sunk data is specified
+    gen = Sink(None, 'result')
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: Could just use keyword argument without explicit None for the first positional argument

File: datalad/crawler/nodes/misc.py (hunk 1 above)
Comment (yarikoptic) [inline datalad/crawler/nodes/misc.py]: And then will come the challenge! Read about dict comprehensions in python, to see how you could do all needed here in 1 line ;-)

File: datalad/crawler/nodes/tests/test_misc.py [hunk 5]
```diff
@@ -47,6 +48,29 @@ def test_get_disposition_filename():
...
+    gen = Sink(output='result')
+    eq_(list(gen(data)), [{'a': 'b', 'x': 'y', 'result': [{'a': 'b', 'x': 'y', 'g': 'h'}], 'g': 'h'}])
+
+    # if list of keys is specified
+    genkeys = Sink(keys, 'result')
+    eq_(list(genkeys(data)), [{'a': 'b', 'x': 'y', 'result': [{'a': 'b', 'x': 'y'}], 'g': 'h'}])
+
+    # if list of keys has no match
+    gentwo = Sink(nomatch, 'result')
+    eq_(list(gentwo(data)), [{'a': 'b', 'result': [{}], 'x': 'y', 'g': 'h'}])
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: oh -- one more test please -- so if you call the same sink second time, it does append new data to the previously collected one.  Also check that actual .data of the sink does contain those entries (above tests tested only output of the call but not the internal state, i.e. `.data` of the sink).

File: datalad/crawler/nodes/misc.py [hunk 6]
```diff
@@ -46,26 +46,29 @@ def __init__(self, keys=None, output=None, ignore_prefixes=['datalad_']):
...
           Keys with which prefixes to ignore.  By default all 'datalad_' ignored
+
         """
         self.data = []
         self.keys = keys
         self.output = output
         self.ignore_prefixes = ignore_prefixes or []
 
-    def get_values(self, *keys):
+    def get_values(self, keys):       # need to fix relevant usages into lists!!
```
Comment (yarikoptic) [inline datalad/crawler/nodes/misc.py]: using TODO:  in your comments would make pycharm "mark" that line which would bring your attention

File: datalad/crawler/nodes/tests/test_misc.py [hunk 7]
```diff
@@ -41,12 +44,62 @@
 @skip_if_no_network
 @use_cassette('brain-map.org-1', return_body='')
 def test_get_disposition_filename():
-    input = {'url': 'http://human.brain-map.org/api/v2/well_known_file_download/157722290'}
-    output = list(get_disposition_filename(input))
+    inputted = {'url': 'http://human.brain-map.org/api/v2/well_known_file_download/157722290'}
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: `inputted` reads awkward -- better use original `input` (and `output`)

File: datalad/crawler/nodes/tests/test_misc.py [hunk 8]
```diff
@@ -41,12 +44,62 @@
...
+    output = list(get_disposition_filename(inputted))
     eq_(len(output), 1)
     eq_(output[0]['filename'], 'T1.nii.gz')
 
 
+def test_get_url_filename():
+    inputted = {'url': 'http://human.brain-map.org/api/v2/well_known_file_download/157722290'}
+    output = list(get_url_filename(inputted))
+    eq_(len(output[0]), 2)
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: why not to check equality of the full output value?

File: datalad/crawler/nodes/misc.py (hunk 6 above)
Comment (yarikoptic) [inline datalad/crawler/nodes/misc.py]: and I believe you already did it, so comment can go away

File: datalad/crawler/nodes/tests/test_misc.py [hunk 9]
```diff
@@ -62,6 +115,28 @@ def test_assign():
     eq_(list(gen(datadup)), [{'x': 'value', 'g': 'y z'}])
 
 
+def test__act_if():
+    values = {'x': 'y', 'a': 'b'}
+    data = {'x': 'y'}
+    datamatch = {'x': 'y', 'a': 'b'}
+
+    # matched = false
+    gen = _act_if(values, False, True)
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: to not bother matching False, True to `re` and `negate` -- use them as keyword args here.  Also test with "default" values (i.e. your False, False below) - i.e. not providing any value here

Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: btw -- `_act_if` is not to be used directly -- it is a base class for `interrupt_if` and `skip_if` and theoretically if any match should raise NotImplementedError from its `_act`.  (eventually -- look into abc module and how to use `@abstractmethod` which should ideally be done for `_act_if._act`

So if testing `_act_if`, tests should raise NotImplementedError when "appropriate"

File: datalad/crawler/nodes/tests/test_misc.py (hunk 7 above)
Comment (glalteva) [inline datalad/crawler/nodes/tests/test_misc.py]: `input` becomes underlined in red for being a built-in name to explain my edit. I switched it back though!

File: datalad/crawler/nodes/tests/test_misc.py (hunk 8 above)
Comment (glalteva) [inline datalad/crawler/nodes/tests/test_misc.py]: My intention was to mimic  `test_get_disposition_filename`. Changed to check equality of full output value.

File: datalad/crawler/nodes/tests/test_misc.py [hunk 10]
```diff
@@ -62,6 +115,51 @@ def test_assign():
...
+    gen = interrupt_if(genact)
+    assert_raises(FinishPipeline, gen._act, data)
+
+
+def test_skip_if():
+    values = {'x': 'y', 'a': 'b'}
+    data = {'x': 'y', 'a': 'b'}
+
+    genact = _act_if(values, re=True, negate=True)
+    gen = skip_if(genact)
```
Comment (yarikoptic) [inline datalad/crawler/nodes/tests/test_misc.py]: this is how should it be
```python
eq_(list(skip_if(data)(data)), [])   # if data and values match precisely
eq_(list(skip_if(data)(updated(data, {'y':2})), [])   # if data has some additional fields -- still matching
eq_(list(skip_if(updated(data, {'y':2}))(data), [data])  # if values to match have don't match all in the data

eq_(list(skip_if(data, negate=True)(data)), [data])   # ...
...
```

**Pull request merged**